Mention that `tarball-ttl` affects global flake registry

Gabriella439 opened the pull request:
This updates the documentation for `tarball-ttl` to mention that global flake registry references (e.g. `nixpkgs#hello`) respect the `tarball-ttl` setting when determining whether to lock the reference again.

# Checklist for maintainers

<!-- Contributors: please leave this as is -->

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

Commit:
Mention that `tarball-ttl` affects global flake registry

This updates the documentation for `tarball-ttl` to mention that global
flake registry references (e.g. `nixpkgs#hello`) respect the
`tarball-ttl` setting when determining whether to lock the reference
again.

edolstra commented on src/libstore/globals.hh:
```diff
@@ -629,6 +629,11 @@ public:
+          This option also affects how often the global flake registry is
+          updated.  For example, a flake URI like `nixpkgs#hello` will trigger
+          a new download of Nixpkgs if the last download is older than the
+          `tarball-ttl`.
```
That's two separate things: checking for a new registry version, and checking whether `github:NixOS/nixpkgs` has a new version. But both are affected by the TTL.

roberth replied:
I didn't catch that confusion. I think this might happen more often as people will think of the flake registry as a typical package registry, whereas it's just a name registry.

Gabriella439 replied:
It's not clear to me what change is being suggested here

roberth replied:
The global flake registry at github.com/NixOS/flake-registry is a downloaded resource just like anything else that's fetched (but not a fixed output derivation). It's a feature that's unrelated to the caching of such downloaded resources, except for the fact that the registry itself is cached. In other words, the registry is not the mechanism that implements the caching.

As such, the second sentence is not an example of the first, despite both being true.

OT: Should we even call it the "global" registry? I think it's too vague, as some tools, notably git, consider the user configuration (of all things) to be "global".

edolstra replied:
The concrete implication is that the tarball TTL doesn't just affect `nix run nixpkgs#hello`, but also `nix run github:NixOS/nixpkgs#hello` (which doesn't use the registry at all).

The registry rarely changes, and certainly what `nixpkgs` refers to doesn't change. So the effect of the TTL on registry caching is usually irrelevant. The new download of Nixpkgs is nt the result of the registry being updated, but the `github:NixOS/nixpkgs` flake being updated.

edolstra replied:
Maybe we should present this as a list of all things affected by the TTL, e.g.

* Files downloaded by `fetchTarball`, etc.
* The registry.
* What revision Git/GitHub"refs" refer to.
* ...

roberth commented on src/libstore/globals.hh:
```diff
@@ -629,6 +629,11 @@ public:
           Files fetched via `NIX_PATH`, `fetchGit`, `fetchMercurial`,
           `fetchTarball`, and `fetchurl` respect this TTL.
+
+          This option also affects how often the global flake registry is
+          updated.  For example, a flake URI like `nixpkgs#hello` will trigger
+          a new download of Nixpkgs if the last download is older than the
+          `tarball-ttl`.
```
```suggestion
          This option also affects how often the global flake registry is
          updated, and how often individual flake inputs are updated.
          For example, a flake URI like `nixpkgs#hello` will trigger
          a new download of Nixpkgs if the last download is older than the
          `tarball-ttl`.
```

I don't think we need an example of the flake registry updating, because that thing doesn't update very often anyway. Analogy with DNS holds up.

OT: Should we add the analogy with DNS to the flake registry docs?

Ericson2314 commented:
This is somewhat off topic, but I see that this setting is only used in `libfetchers` and downstream. If it could be moved to `FetchSettings` that would be much appreciated!

Gabriella439 committed:
Address review feedback

… by explaining the distinction between the various types of
flake reference updates

Gabriella439 commented:
Alright, I think I addressed the review feedback

tomberek commented:
stale: re-open and re-base as needed
